examples/small-coalton-programs/src/brainfold.lisp

```diff
@@ -0,0 +1,308 @@
+;;;; (run-file "/path/to/your/file.bf")
+;;;;
+;;;; try (coalton (hello-world)) in the REPL!
+;;;;
+
+(cl:defpackage #:brainfold
+  (:use
+   #:coalton
+   #:coalton-prelude)
+  (:local-nicknames
+   (#:vec #:coalton-library/vector)
+   (#:iter #:coalton-library/iterator)
+   (#:cell #:coalton-library/cell)
+   (#:char #:coalton-library/char)
+   (#:str #:coalton-library/string)
+   (#:list #:coalton-library/list)
+   (#:arith #:coalton-library/math))
+
+  (:export
+   #:eval
+   #:run-program
+   #:run-file
+
+   ;; Examples
+   #:hello-world
+   #:gnarly-hello-world
+   #:squares))
+
+(in-package #:brainfold)
+
+(named-readtables:in-readtable coalton:coalton)
+
+;;;
+;;; State/Env
+;;;
+
+(coalton-toplevel
+
+  (define-type State
+    (State (Vector Integer) (Cell UFix) (Cell String))) ; Memory, Pointer, Print-buffer
+
+  (define (state-memory (State memory _ _))
+    "Accessor for the State memory."
+    memory)
+  
+  (define (state-pointer (State _ pointer _))
+    "Accessor for the State pointer."
+    pointer)
+
+  (define (state-print-buffer (State _ _ buffer))
+    "Accessor for the State buffer."
+    buffer)
+
+  ;;
+  ;; Generating a Brainfold memory vector
+  ;;
+
+  (declare bf-vector-size UFix)
+  (define bf-vector-size 1000)
+  
+  (define (generate-bf-vector)
+    "Initializes the brainfold array."
+    (let v = (vec:new))
+    (vec:extend!  v (iter:repeat-for 0 bf-vector-size))
+    v)
+
+  ;;
+  ;; Generating a default State:
+  ;;
+
+  (define (new-state)
+    "Generates a new, blank Brainfold State."
+    (State (generate-bf-vector) (cell:new 0) (cell:new "")))
+
+  ;;
+  ;; Accessing the current value
+  ;;
+
+  (declare value-at-pointer (State -> Integer))
+  (define (value-at-pointer (State memory pointer _))
+    "Returns the value at the current pointer."
+    (unwrap (vec:index (cell:read pointer)
+		       memory))))
+
+
+;;;
+;;; Commands (Functions called by Brainfold Cmds)
+;;;
+
+(coalton-toplevel
+
+  ;;
+  ;; Navigating through bf-cells (> <)
+  ;;
+  
+  (declare move-right (State -> State))
+  (define (move-right s)
+    "Moves the pointer one bf-cell to the right."
+    (cell:increment! (state-pointer s))
+    s)
+
+  (declare move-left (State -> State))
+  (define (move-left s)
+    "Moves the pointer one bf-cell to the left."
+    (cell:decrement! (state-pointer s))
+    s)
+
+  ;;
+  ;; Changing bf-cell values (+ -)
+  ;;
+
+  (declare incr (State -> State))
+  (define (incr s)
+    "Increments the value for the current bf-cell."
+    (vec:set! (cell:read (state-pointer s))
+              (1+ (value-at-pointer s))
+              (state-memory s))
+    s)
+
+  (declare decr (State -> State))
+  (define (decr s)
+    "Decrements the value for the current bf-cell."
+    (vec:set! (cell:read (state-pointer s))
+              (1- (value-at-pointer s))
+              (state-memory s))
+    s)
+
+  ;;
+  ;; Printing Cells (.)
+  ;;
+  
+  (declare find-ascii (UFix -> String))
+  (define (find-ascii ascii-value)
```

Comment:
ascii->string

Reply:
Should I define an Ascii type?

Reply:
No, I think ASCII as being a byte representation is fine. Technically U8 would be "more correct" but it's no big deal.

Reply:
I changed it, right hammer for the job

Reply:
I changed it and changed it back- `char-code` and `code-char` both operate with UFix.